### PR TITLE
[None][feat] DSv4 prep: MoE routing and backend support

### PR DESCRIPTION
**Pull request by lfr-0531**
## Release Notes

* **New Features**
  * Added support for DeepSeekV4 model with new MoE routing method and gate-forward operation
  * Introduced optional SwigLU clamping support for FP8 quantization paths
  * Added MPI-based distributed initialization bootstrap for MegaMoE under MPI executor

* **Improvements**
  * Enhanced MoE routing methods to support token-aware routing via input IDs
  * Improved FP8 activation kernel with configurable uniform clamp limits
  * Extended Cutlass kernel heuristics for optimized SM100/SM103 performance

* **Tests**
  * Added comprehensive test suite for new MoE gate operation
  * New routing regression tests for DeepSeekV3/V4 configuration validation

## Description

This is part of the DSv4 split from umbrella PR #14751. It isolates shared MoE/routing backend support so it can be reviewed independently from the DSv4 model and sparse attention wiring.

Included here:
- DeepGEMM dependency update used by MoE backends.
- MoE routing kernels and thop bindings, including `gate_forward`, `dsv3_router_gemm_op`, and `RoutingKernelTopK.cuh`.
- Block-scale MoE runner updates, DeepGEMM/MegaMoE plumbing, routing scheduler changes, `swiglu_limit` plumbing, and DeepSeekV4 routing interface changes.
- CUTLASS MoE candidate filtering needed for `CUTLASS + NVFP4` on SM100-family fallback.

This PR has no dependency on earlier DSv4 split PRs after current `main`. PR-9 will depend on this PR for the final DSv4 model/API/tokenizer wiring.

Intentionally excluded: DSv4 model/config/tokenizer files, sparse attention backend/cache-manager files, compressor/mHC implementation, IndexerTopK implementation, and attention fusion custom ops.

Risk: this touches shared MoE backend code used by non-DSv4 consumers. The default-path behavior is covered by existing MoE module/routing tests below.

## Test Coverage

Build/install:
- `CCACHE_DIR=/home/scratch.fanrongl_coreai/ccache_trtllm python3 ./scripts/build_wheel.py --trt_root /usr/local/tensorrt --benchmarks --use_ccache --cuda_architectures "90-real;100-real" --configure_cmake`
- `./.venv-3.12/bin/pip install --force-reinstall --no-deps build/tensorrt_llm-1.3.0rc19-cp312-cp312-linux_x86_64.whl`

Smoke/static checks:
- Import smoke confirmed `tensorrt_llm.__file__` resolves to this PR worktree.
- Verified `fused_moe`, `moe_alltoall.py`, `torch.ops.trtllm.gate_forward`, `torch.ops.trtllm.dsv3_router_gemm_op`, `torch.ops.trtllm.fp8_block_scale_moe_runner`, and `tensorrt_llm.deep_gemm_cpp_tllm`.
- Routing signature smoke: `DeepSeekV4MoeRoutingMethod.apply(..., input_ids=...)` accepts optional `input_ids`; `Gemma4MoeRoutingMethod.apply` is restored to `apply(router_logits)` and PR-8 no longer changes `modeling_gemma4.py`.
- `git diff --check github/main..HEAD`
- PR-8 scope-leakage grep for attention backend, DSv4 model/tokenizer, compressor/mHC, IndexerTopK, and attention-fusion files.
- `./.venv-3.12/bin/pre-commit run --files $(git diff --name-only HEAD)` before the final rebase.

GPU tests after checking `nvidia-smi`; used idle `CUDA_VISIBLE_DEVICES=1`:
- `timeout 25m ./.venv-3.12/bin/python -m pytest tests/unittest/_torch/modules/moe/test_moe_module.py tests/unittest/_torch/thop/serial/test_moe_gate.py` -> 270 passed, 2116 skipped.
- `timeout 20m ./.venv-3.12/bin/python -m pytest tests/unittest/_torch/modules/test_moe_routing.py tests/unittest/_torch/modules/test_moe_load_balancer.py` -> 527 passed, 87 skipped.
- `timeout 20m ./.venv-3.12/bin/python -m pytest tests/unittest/_torch/thop/parallel/test_dsv3_router_gemm.py` -> 48 passed.

After rebasing onto latest `main` (`0ec325040087af5f98d040fe1dcf59273c19d209`), new main only touched `tests/integration/defs/perf/utils.py` and did not overlap PR-8 files. Rebase verification:
- import/routing signature smoke passed after the amend; `modeling_gemma4.py` is absent from `github/main..HEAD`.
- `git diff --check github/main..HEAD` passed.
- scope-leakage grep passed.
- `timeout 20m ./.venv-3.12/bin/python -m pytest tests/unittest/_torch/modules/test_moe_routing.py tests/unittest/_torch/modules/test_moe_load_balancer.py tests/unittest/_torch/thop/parallel/test_dsv3_router_gemm.py` -> 575 passed, 87 skipped.
- Post-amend rerun on `CUDA_VISIBLE_DEVICES=1`: `timeout 45m ./.venv-3.12/bin/python -m pytest -q tests/unittest/_torch/modules/test_moe_routing.py tests/unittest/_torch/modules/test_moe_load_balancer.py tests/unittest/_torch/thop/parallel/test_dsv3_router_gemm.py` -> 575 passed, 87 skipped.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tests/unittest/_torch/modules/moe/test_moe_module.py [hunk 1]
```diff
@@ -1307,6 +1307,20 @@ def generate_base_test_params(
 )
 
 
+def _skip_cutlass_nvfp4_sm120(
```
Comment (xxi-nv) [inline tests/unittest/_torch/modules/moe/test_moe_module.py]: Please help to move this function into should_skip_cutlass, instead of creating a seperate function.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py [hunk 2]
```diff
@@ -1139,6 +1138,7 @@ def forward_chunk(
         self,
         x: Union[torch.Tensor, Fp4QuantizedTensor],
         router_logits: torch.Tensor,
+        input_ids: Optional[torch.IntTensor],
```
Comment (xxi-nv) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py:1144]: Suggestion: confine the new input_ids plumbing to moe_scheduler.py instead of threading it through every backend and routing method.

input_ids is consumed in exactly one place: DeepSeekV4MoeRoutingMethod.apply (hashed gate). For the ConfigurableMoE path, routing_method.apply(...) is only ever called from the scheduler (_forward_chunk_impl / the fused _forward_chunk). So input_ids only needs to be (a) received by the scheduler and (b) sliced in lockstep with x / router_logits where the scheduler already chunks them.

As written, this PR also adds an input_ids parameter (and a parallel .split()) to forward_impl / forward_chunk of every backend (fused_moe_cutlass, fused_moe_deepgemm, fused_moe_cute_dsl, fused_moe_triton, fused_moe_vanilla, fused_moe_wide_ep) and to ~10 routing-method signatures. That is a large API surface change for a feature used by only the DSv4 hashed gate. Two reasons it is avoidable:

Those backend.forward_chunk routing calls are the legacy standalone path; on the ConfigurableMoE path routing is owned by the scheduler, not the backend.
When a backend routes in-kernel (router_logits passed to the kernel, e.g. TEP without DP comm), routing_method.apply is not called at all, so input_ids is meaningless on that path.
Proposal: keep input_ids only in moe_scheduler.py (plus BaseMoeRoutingMethod.apply getting a default and DeepSeekV4MoeRoutingMethod reading it), and revert it from the per-backend forward_chunk / forward_impl signatures. This also matches the design rule that forward-time routing policy lives in the scheduler.

**Comment by xxi-nv** [comment]
Suggestion: collapse swiglu_limit_scalar into swiglu_limit; don't add it to backends that consume the per-expert tensor.

swiglu_limit (per-expert tensor) and swiglu_limit_scalar (uniform float) are the same concept (the SwiGLU clamp limit) in two representations. The split leaks an internal kernel-ABI difference into the public module/backend constructors, and the scalar is plumbed into backends that never read it.

Who actually consumes the scalar — only the FP8 separate-activation path:

fused_moe_deepgemm.py:726 → silu_and_mul_masked_post_quant_fwd(swiglu_limit=self.swiglu_limit_scalar)
ops/moe_op_deepgemm.py:266 → swiglu_limit=getattr(module, "swiglu_limit_scalar", None)
fused_moe_trtllm_gen.py:684 → gemm1_clamp_limit=self.swiglu_limit_scalar
mega_moe_deepgemm.py:731 → activation_clamp=self.swiglu_limit_scalar
Who does NOT consume it but still gets the new ctor arg: CutlassFusedMoE (uses the tensor via moe_op_cutlass.py:166), WideEPMoE (stores only), Triton (uses the tensor). For those, swiglu_limit_scalar is a dead parameter.

Why one public param is enough: the model always provides a uniform clamp (DSv4 config.swiglu_limit = 10.0). The per-expert tensor is derived internally for NVFP4 only, because the clamp is folded with per-expert fc31_alpha (quantization.py:3954 swiglu_limit.data.div_(fc31_alpha)) — that division is what makes it genuinely per-expert, not the public API. And the tensor→scalar reduction already exists: mega_moe_cute_dsl.py:_resolve_gate_up_clamp takes the per-layer swiglu_limit tensor, asserts uniformity, and reduces to one float.

Proposed shape:

Keep a single public swiglu_limit: Union[float, torch.Tensor] (or just float, since both known consumers — DSv4 and GPT-OSS — use a uniform limit). Drop swiglu_limit_scalar from MoE.__init__ and from every backend constructor.
Each backend derives the form its kernel needs, internally:
NVFP4/MXFP4 fused-activation: keep the existing ÷ fc31_alpha post-load step → per-expert tensor (internal detail).
FP8 separate-activation: use the float directly (reuse the _resolve_gate_up_clamp uniformity-checked reduction when the input is a tensor).
This removes the dead ctor args on Cutlass/WideEP/Triton and keeps the per-expert representation as an internal NVFP4 weight-post-load concern rather than a second user-facing parameter.

**For the MOE API part, I will do the refactor in <html>
<body>

TRTLLM-13501
--



</body>
</html>[TRTLLM-13501](https://jirasw.nvidia.com/browse/TRTLLM-13501)**

**Review by xxi-nv — APPROVED** [review]
To unblock the merge, For the MOE API part, I will do the refactor in 
TRTLLM-13501 -- [TRTLLM-13501](https://jirasw.nvidia.com/browse/TRTLLM-13501)

**Review by hyukn — APPROVED** [review]

**Comment by xxi-nv** [comment]
@sunnyqgg @rosenrodt Please help to reivew the code change in TRTLLMGEN.

**Comment by xxi-nv** [comment]
/bot run --stage-list "A100X-PyTorch-1, DGX_B200-PyTorch-5, B300-PyTorch-1, DGX_B200-PyTorch-3"

File: tensorrt_llm/_torch/modules/fused_moe/routing.py [hunk 3]
```diff
@@ -471,8 +496,104 @@ def e_score_correction_bias(self) -> torch.Tensor:
...
+        super().__init__()
+        self._top_k = top_k
+        self.n_group = n_group
+        self.topk_group = topk_group
+        self.routed_scaling_factor = routed_scaling_factor
+        self.is_hashed = is_hashed
+
+        # Pass a callable to fetch the tensor from DeepseekV4Gate at runtime, ensuring it is on the correct device
+        assert callable(callable_e_score_correction_bias)
+        assert callable(callable_tid2eid)
```
Comment (syuoni) [inline tensorrt_llm/_torch/modules/fused_moe/routing.py:533]: Hi, may I ask why do we need these two callables instead of direct tensors/Parameters? I suppose the later also handles device correctly.

File: tensorrt_llm/_torch/modules/fused_moe/routing.py [hunk 4]
```diff
@@ -471,8 +496,104 @@ def e_score_correction_bias(self) -> torch.Tensor:
...
+                                           device=logits.device,
+                                           requires_grad=False)
+            tid2eid_tensor = torch.empty(0,
+                                         dtype=torch.int32,
+                                         device=logits.device,
+                                         requires_grad=False)
+            torch.ops.trtllm.gate_forward(
+                logits, self.callable_e_score_correction_bias(),
+                input_ids_tensor, tid2eid_tensor, out_weights, out_indices,
+                self._top_k, self.routed_scaling_factor, False)
```
Comment (syuoni) [inline tensorrt_llm/_torch/modules/fused_moe/routing.py:578]: Seems `is_hash` is always associated with valid `input_ids` and `tid2eid`. 

* For the `is_hash=False` path, can we let `torch.ops.trtllm.gate_forward` accept `tid2eid: Optional[Tensor]`? So, we can infer that `is_hash=False` if `tid2eid is None` and we can remove the explicit `is_hash` arg?
* For the `is_hash=True` path, similarly we can make `bias: Optional[Tensor]`.

**Review by syuoni — APPROVED** [review]
LGTM, left some minor comments.

File: tensorrt_llm/_torch/modules/fused_moe/routing.py (hunk 4 above)
Comment (xxi-nv) [reply tensorrt_llm/_torch/modules/fused_moe/routing.py:578]: Thanks for the suggestion. I believe you're right, given that this is a prerequisite for the following model PR. I'd prefer to leave it as it is and refactor once the DS - V4 model has also been merged back into MAIN.
Otherwise, we have to change the API.

File: tensorrt_llm/_torch/modules/fused_moe/routing.py (hunk 3 above)
Comment (xxi-nv) [reply tensorrt_llm/_torch/modules/fused_moe/routing.py:533]: It adheres to the current mechanism in the DeepSeekV3MoeRoutingMethod.

**Review by yuanjingx87 — APPROVED** [review]
approve on oss-compliance perspective.
